File: adapters/python/src/writ_agent/claude_agent_sdk.py
"""Claude Agent SDK integration (package ``claude-agent-sdk``).

Integration point: SDK hook callbacks registered through
``ClaudeAgentOptions(hooks=...)``:

* ``PreToolUse`` -> ``decide`` (+ approver + ``resolve`` for deferred asks).
  Answers ``permissionDecision: "allow"`` or ``"deny"`` with writ's reason.
* ``PostToolUse`` -> ``complete``. For a redact verdict it returns
  ``updatedToolOutput`` with the redacted result.
* ``PostToolUseFailure`` -> ``complete(ok=False)``.

Why hooks and not ``can_use_tool``: the CLI consults ``can_use_tool`` only
when its own permission rules would prompt; tools already allowed by
``allowed_tools`` / ``permission_mode`` never reach it, and it has no
post-execution step for ``complete`` or redaction. A ``PreToolUse`` hook
with no matcher runs for every tool call.

Fail closed: the callbacks never raise (an exception in an SDK hook becomes
an error reply whose handling is up to the CLI). Every failure answers
``deny`` in ``PreToolUse``; in ``PostToolUse`` a redacted call whose
redaction cannot be completed has every string in its output masked.

Tool names and inputs are mapped with :func:`writ_agent.toolmap.map_claude_tool`,
the same table ``writ check --format claude-code`` uses.
"""

from __future__ import annotations

import json
import re
from typing import Any, Mapping

from claude_agent_sdk import HookMatcher

from .errors import WritError
from .guard import DEFAULT_APPROVAL_TIMEOUT, Writ, refusal_text, to_text
from .toolmap import map_claude_tool
from .types import Caller, Decision

try:  # pragma: no cover
    from importlib.metadata import version as _pkg_version

    _SDK_VERSION: str | None = _pkg_version("claude-agent-sdk")
except Exception:  # pragma: no cover
    _SDK_VERSION = None

MASK = "[redacted-by-writ]"
_EXIT_RE = re.compile(r"Exit code (\d+)")


def _mask_all(v: Any) -> Any:
    if isinstance(v, str):
        return MASK
    if isinstance(v, list):
        return [_mask_all(x) for x in v]
    if isinstance(v, dict):
        return {k: _mask_all(x) for k, x in v.items()}
    return v


def _same_shape(a: Any, b: Any) -> bool:
    if isinstance(a, dict):
        return isinstance(b, dict) and a.keys() == b.keys() and all(_same_shape(a[k], b[k]) for k in a)
    if isinstance(a, list):
        return isinstance(b, list) and len(a) == len(b) and all(_same_shape(x, y) for x, y in zip(a, b))
    if isinstance(a, str):
        return isinstance(b, str)
    return type(a) is type(b) and a == b


def _pre(decision: str, reason: str | None = None) -> dict[str, Any]:
    out: dict[str, Any] = {"hookEventName": "PreToolUse", "permissionDecision": decision}
    if reason:
        out["permissionDecisionReason"] = reason
    return {"hookSpecificOutput": out}


class WritClaudeHooks:
    """writ hooks for ``ClaudeAgentOptions(hooks=WritClaudeHooks(writ).hooks())``.

    Args:
        writ: the :class:`~writ_agent.Writ` gate. Give it an ``approver`` to
            handle ``ask`` verdicts (the client then runs ``--ask defer``);
            without one, asks are rejected.
        session_id: override the SDK's session id.
        mcp_transports: ``{server name: "stdio"|"sse"|"http"|"sdk"}`` for the
            ``server.transport`` writ records (default ``"unknown"``).
        allow_decision: what a writ allow answers in ``PreToolUse``:
            ``"allow"`` (default, as ``writ check --format claude-code`` does)
            skips the SDK's own permission prompt; ``None`` returns no
            decision so the SDK's permission rules still apply on top.
        hook_timeout: seconds the CLI waits for the ``PreToolUse`` hook.
            Default covers the gateway timeout plus the approval timeout.
    """

    def __init__(
        self,
        writ: Writ,
        *,
        session_id: str | None = None,
        mcp_transports: Mapping[str, str] | None = None,
        allow_decision: str | None = "allow",
        hook_timeout: float | None = None,
    ) -> None:
        if allow_decision not in ("allow", None):
            raise ValueError("allow_decision must be 'allow' or None")
        self.writ = writ
        self.session_id = session_id
        self.mcp_transports = dict(mcp_transports or {})
        self.allow_decision = allow_decision
        self.hook_timeout = hook_timeout
        self._pending: dict[tuple[str, str], Decision] = {}

    # -- helpers -----------------------------------------------------------

    def _caller(self, inp: Mapping[str, Any]) -> Caller:
        c = self.writ.caller
        if c.agent == "writ-agent":
            agent_id = inp.get("agent_id")
            return Caller(
                agent="claude-agent-sdk",
                agent_version=_SDK_VERSION,
                user=c.user,
                non_human_id=str(agent_id) if agent_id else c.non_human_id,
            )
        return c

    def _key(self, inp: Mapping[str, Any], tool_use_id: str | None) -> tuple[str, str]:
        session = self.session_id or str(inp.get("session_id") or self.writ.session_id)
        return session, str(inp.get("tool_use_id") or tool_use_id or "")

    # -- hooks -------------------------------------------------------------

    async def pre_tool_use(self, input: Any, tool_use_id: str | None, context: Any) -> dict[str, Any]:
        try:
            inp: Mapping[str, Any] = input if isinstance(input, Mapping) else {}
            session, use_id = self._key(inp, tool_use_id)
            name = inp.get("tool_name")
            if not isinstance(name, str) or not name:
                return _pre("deny", "writ: hook input has no tool_name (fail closed)")
            tool_input = inp.get("tool_input")
            mapped = map_claude_tool(
                name, tool_input if isinstance(tool_input, Mapping) else {}, mcp_transports=self.mcp_transports
            )
            call = self.writ.make_call(
                mapped.tool,
                mapped.args,
                session_id=session,
                call_id=use_id or None,
                server=mapped.server,
                caller=self._caller(inp),
            )
            d = await self.writ.aauthorize(call)
            if use_id:
                self._pending[(session, use_id)] = d
            if self.allow_decision is None:
                return {}
            return _pre("allow", f"writ {d.decision}" + (f" (rule '{d.rule_id}')" if d.rule_id else ""))
        except WritError as e:
            return _pre("deny", refusal_text(e))
        except Exception as e:  # noqa: BLE001 - any adapter bug must deny, not raise
            return _pre("deny", f"writ adapter error ({type(e).__name__}: {e}); the tool did not run")

    async def post_tool_use(self, input: Any, tool_use_id: str | None, context: Any) -> dict[str, Any]:
        inp: Mapping[str, Any] = input if isinstance(input, Mapping) else {}
        d = self._pending.pop(self._key(inp, tool_use_id), None)
        if d is None:
            return {}  # no decision of ours (e.g. PreToolUse denied elsewhere)
        response = inp.get("tool_response")
        try:
            redacted = await self.writ.arecord(d, ok=True, output=to_text(response))
        except WritError as e:
            if not d.is_redact:
                return {"systemMessage": f"writ could not record this tool execution: {e}"}
            redacted = None
        except Exception:  # noqa: BLE001
            if not d.is_redact:
                return {}
            redacted = None
        if not d.is_redact:
            return {}
        return {
            "hookSpecificOutput": {
                "hookEventName": "PostToolUse",
                "updatedToolOutput": self._redacted_output(response, redacted),
            }
        }

    @staticmethod
    def _redacted_output(original: Any, redacted: str | None) -> Any:
        """Shape-preserving redacted tool_response; masks everything on doubt."""
        if redacted is None:
            return _mask_all(original)
        if isinstance(original, str):
            return redacted
        try:
            parsed = json.loads(redacted)
        except ValueError:
            return _mask_all(original)
        # Masking may only change string contents. Anything else means a
        # pattern hit JSON syntax, and the result is not trusted.
        return parsed if _same_shape(original, parsed) else _mask_all(original)

    async def post_tool_use_failure(self, input: Any, tool_use_id: str | None, context: Any) -> dict[str, Any]:
        inp: Mapping[str, Any] = input if isinstance(input, Mapping) else {}
        d = self._pending.pop(self._key(inp, tool_use_id), None)
        if d is None:
            return {}
        err = str(inp.get("error") or "")
        m = _EXIT_RE.search(err)
        try:
            await self.writ.arecord(d, ok=False, output=err, exit=int(m.group(1)) if m else 1)
        except Exception:  # noqa: BLE001 - nothing to withhold; the tool already failed
            pass
        return {}

    # -- wiring ------------------------------------------------------------

    def _pre_timeout(self) -> float:
        if self.hook_timeout is not None:
            return self.hook_timeout
        approval = self.writ.approval_timeout or DEFAULT_APPROVAL_TIMEOUT
        return 2 * self.writ.client.timeout + (approval if self.writ.client.ask == "defer" else 0) + 5

    def hooks(self) -> dict[str, list[HookMatcher]]:
        """The ``hooks=`` mapping for ``ClaudeAgentOptions``; merge with your own if needed."""
        post_timeout = self.writ.client.timeout + 5
        return {
            "PreToolUse": [HookMatcher(matcher=None, hooks=[self.pre_tool_use], timeout=self._pre_timeout())],
            "PostToolUse": [HookMatcher(matcher=None, hooks=[self.post_tool_use], timeout=post_timeout)],
            "PostToolUseFailure": [
                HookMatcher(matcher=None, hooks=[self.post_tool_use_failure], timeout=post_timeout)
            ],
        }


def writ_hooks(writ: Writ, **kwargs: Any) -> dict[str, list[HookMatcher]]:
    """Shorthand: ``ClaudeAgentOptions(hooks=writ_hooks(writ))``."""
    return WritClaudeHooks(writ, **kwargs).hooks()


__all__ = ["WritClaudeHooks", "writ_hooks"]

File: adapters/typescript/src/claude-agent-sdk.ts
/**
 * writ integration for the Claude Agent SDK (`@anthropic-ai/claude-agent-sdk`).
 *
 * Plugs into `query({ options: { hooks, canUseTool } })`:
 * - `PreToolUse` asks writ for a verdict and returns the SDK's
 *   `permissionDecision` (`allow` / `deny` / `ask`).
 * - `PostToolUse` / `PostToolUseFailure` record the execution (`complete`)
 *   and, for a redact verdict, replace the tool output via `updatedToolOutput`.
 * - `canUseTool` resolves writ asks that were routed to the SDK's own
 *   permission flow.
 *
 * Only SDK *types* are imported; this module has no runtime dependency on the SDK.
 */
import type {
  CanUseTool,
  HookCallback,
  HookCallbackMatcher,
  HookEvent,
  HookInput,
  PermissionResult,
  PostToolUseFailureHookInput,
  PostToolUseHookInput,
  PreToolUseHookInput,
  SyncHookJSONOutput,
} from "@anthropic-ai/claude-agent-sdk";

import { shouldDispatch, type Approver, type WritClient } from "./client.js";
import { describeBlock, WritError } from "./errors.js";
import { fromRedacted, outputText, WITHHELD_OUTPUT } from "./output.js";
import type { CallerIdentity, Decision, ServerIdentity, ToolCallInput } from "./protocol.js";

/** A Claude tool use normalized to writ's vocabulary. */
export interface MappedToolCall {
  tool: string;
  args: Record<string, unknown>;
  server?: ServerIdentity | null;
}

/** MCP server provenance as the SDK reports it (`mcp_server` / `mcpServer`). */
export interface McpProvenance {
  name: string;
  source: string;
}

const SHELL_TOOLS = new Set(["Bash", "PowerShell"]);
const READ_TOOLS = new Set(["Read", "Glob", "Grep", "LS", "NotebookRead"]);
const WRITE_TOOLS = new Set(["Write", "Edit", "MultiEdit", "NotebookEdit"]);

function asRecord(input: unknown): Record<string, unknown> {
  return typeof input === "object" && input !== null && !Array.isArray(input)
    ? { ...(input as Record<string, unknown>) }
    : { input };
}

function firstString(args: Record<string, unknown>, keys: string[]): string | undefined {
  for (const k of keys) {
    const v = args[k];
    if (typeof v === "string") return v;
  }
  return undefined;
}

/**
 * Map a Claude tool name + input to writ's policy vocabulary. Kept identical
 * to `writ check --format claude-code` (crates/writ-cli/src/hook.rs):
 * - `Bash` / `PowerShell` → `bash` (with `command`)
 * - `Read` / `Glob` / `Grep` / `LS` / `NotebookRead` → `fs.read`; `Write` / `Edit` / `MultiEdit` / `NotebookEdit` → `fs.write` (with `path`)
 * - `WebFetch` → `http` (with `url`); `WebSearch` → `web.search` (with `query`)
 * - `mcp__<server>__<tool>` → tool `<tool>` with `server: { name: <server> }`
 * - anything else keeps its SDK name.
 * The original input fields are kept; normalized keys are added alongside.
 */
export function mapClaudeTool(toolName: string, input: unknown, mcpServer?: McpProvenance): MappedToolCall {
  const args = asRecord(input);
  if (SHELL_TOOLS.has(toolName)) {
    const command = firstString(args, ["command"]);
    return { tool: "bash", args: command === undefined ? args : { ...args, command } };
  }
  if (READ_TOOLS.has(toolName) || WRITE_TOOLS.has(toolName)) {
    // An explicit string `path` wins; otherwise the tool's native key.
    const path = firstString(args, ["path", "file_path", "notebook_path"]);
    const tool = WRITE_TOOLS.has(toolName) ? "fs.write" : "fs.read";
    return { tool, args: path === undefined ? args : { ...args, path } };
  }
  if (toolName === "WebFetch") return { tool: "http", args };
  if (toolName === "WebSearch") return { tool: "web.search", args };
  if (toolName.startsWith("mcp__")) {
    const rest = toolName.slice("mcp__".length);
    const sep = rest.indexOf("__");
    if (sep > 0 && sep < rest.length - 2) {
      const serverName = mcpServer?.name ?? rest.slice(0, sep);
      return {
        tool: rest.slice(sep + 2),
        args,
        server: { name: serverName, transport: "unknown" },
      };
    }
  }
  return { tool: toolName, args };
}

export interface WritClaudeOptions {
  client: WritClient;
  /**
   * Decides deferred asks inline, inside the PreToolUse hook (client must use
   * `ask: "defer"`). Takes precedence over `canUseTool`.
   */
  approver?: Approver;
  /**
   * Your own SDK permission callback. With `ask: "defer"` and no `approver`,
   * writ asks are handed to the SDK's permission flow (`permissionDecision:
   * "ask"`) and this callback decides them; its answer is sent to writ as
   * `resolve`. It is also consulted for SDK permission prompts that writ did
   * not raise. Without it, those are denied.
   */
  canUseTool?: CanUseTool;
  /** Default `{ agent: "claude-agent-sdk" }`. */
  caller?: CallerIdentity;
  /** Override the tool mapping (default `mapClaudeTool`). */
  mapTool?: (toolName: string, input: unknown, mcpServer?: McpProvenance) => MappedToolCall;
  /** Hook matcher (tool name regex); default: every tool. */
  matcher?: string;
  /** SDK hook timeout in seconds (covers inline approvals). */
  hookTimeoutSec?: number;
  approvalTimeoutMs?: number;
  /**
   * What an allow / redact verdict returns to the SDK: `"allow"` (default, as
   * `writ check --format claude-code`) skips the SDK's own permission prompt;
   * `"passthrough"` returns no decision, so the SDK's permission rules still
   * apply on top of writ.
   */
  onAllow?: "allow" | "passthrough";
}

interface Tracked {
  ref: string;
  decision: Decision;
}

const MAX_TRACKED = 10_000;

/** The pieces to spread into the SDK's `query({ options })`. */
export interface WritClaudeIntegration {
  hooks: Partial<Record<HookEvent, HookCallbackMatcher[]>>;
  canUseTool: CanUseTool;
  preToolUse: HookCallback;
  postToolUse: HookCallback;
  postToolUseFailure: HookCallback;
}

function preOutput(permissionDecision: "allow" | "deny" | "ask", reason: string): SyncHookJSONOutput {
  return {
    hookSpecificOutput: {
      hookEventName: "PreToolUse",
      permissionDecision,
      permissionDecisionReason: reason,
    },
  };
}

function allowReason(d: Decision): string {
  const rule = d.rule_id ? `rule '${d.rule_id}'` : "policy default";
  return d.decision === "redact" ? `writ: allowed with redaction by ${rule}` : `writ: allowed by ${rule}`;
}

function askReason(d: Decision, tool: string): string {
  const rule = d.rule_id ? `rule '${d.rule_id}'` : "policy default";
  const where = d.location ? ` (${d.location})` : "";
  const why = d.reason ? `: ${d.reason}` : "";
  return `writ: '${tool}' needs approval by ${rule}${where}${why}`;
}

function errorText(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

function sameJson(a: unknown, b: unknown): boolean {
  try {
    return JSON.stringify(a) === JSON.stringify(b);
  } catch {
    return false;
  }
}

/**
 * Build writ hooks and a `canUseTool` callback for the Claude Agent SDK.
 * Every failure path denies the tool (fail closed): a hook never throws.
 */
export function createWritIntegration(options: WritClaudeOptions): WritClaudeIntegration {
  const { client } = options;
  const caller = options.caller ?? client.caller ?? { agent: "claude-agent-sdk" };
  const mapTool = options.mapTool ?? mapClaudeTool;
  const onAllow = options.onAllow ?? "allow";
  const dispatched = new Map<string, Tracked>();
  const pendingAsks = new Map<string, Tracked>();

  const track = (map: Map<string, Tracked>, key: string, value: Tracked): void => {
    if (map.size >= MAX_TRACKED) {
      const oldest = map.keys().next();
      if (oldest.done !== true) map.delete(oldest.value);
    }
    map.set(key, value);
  };

  const buildCall = (input: PreToolUseHookInput, toolUseID: string | undefined): ToolCallInput => {
    const mapped = mapTool(input.tool_name, input.tool_input, input.mcp_server);
    const call: ToolCallInput = {
      session_id: input.session_id || client.sessionId,
      tool: mapped.tool,
      args: mapped.args,
      // Subagent tool calls carry the subagent id as the non-human identity.
      caller: input.agent_id ? { ...caller, non_human_id: input.agent_id } : caller,
    };
    const id = input.tool_use_id || toolUseID;
    if (id) call.call_id = id;
    if (mapped.server !== undefined) call.server = mapped.server;
    return call;
  };

  const preToolUse: HookCallback = async (input: HookInput, toolUseID, { signal }) => {
    if (input.hook_event_name !== "PreToolUse") return {};
    const key = input.tool_use_id || toolUseID || "";
    try {
      if (signal.aborted) return preOutput("deny", "writ: aborted before a decision (fail closed)");
      const call = buildCall(input, toolUseID);
      let decision: Decision;
      if (options.approver !== undefined) {
        decision = await client.authorize(call, {
          approver: options.approver,
          ...(options.approvalTimeoutMs !== undefined ? { approvalTimeoutMs: options.approvalTimeoutMs } : {}),
        });
      } else {
        decision = await client.decide(call);
        if (decision.decision === "ask" && decision.approval === "required") {
          if (decision.ref === undefined) throw new WritError("protocol", "deferred ask is missing ref");
          if (options.canUseTool !== undefined && key !== "") {
            track(pendingAsks, key, { ref: decision.ref, decision });
            return preOutput("ask", askReason(decision, input.tool_name));
          }
          decision = await client.resolve(decision.ref, false, "adapter:no-approver");
        }
      }
      if (!shouldDispatch(decision) || decision.ref === undefined) {
        return preOutput("deny", describeBlock(decision, input.tool_name));
      }
      if (key === "" && decision.decision === "redact") {
        return preOutput("deny", `writ: redact verdict for '${input.tool_name}' needs a tool_use_id to apply (fail closed)`);
      }
      if (key !== "") track(dispatched, key, { ref: decision.ref, decision });
      return onAllow === "passthrough" ? {} : preOutput("allow", allowReason(decision));
    } catch (err) {
      return preOutput("deny", `writ: cannot authorize '${input.tool_name}' (fail closed): ${errorText(err)}`);
    }
  };

  const postToolUse: HookCallback = async (input: HookInput, toolUseID) => {
    if (input.hook_event_name !== "PostToolUse") return {};
    const post = input as PostToolUseHookInput;
    const key = post.tool_use_id || toolUseID || "";
    const tracked = dispatched.get(key);
    if (tracked === undefined) {
      const unresolved = pendingAsks.get(key);
      if (unresolved !== undefined) {
        // The tool ran although its writ ask was never resolved: record the
        // rejection and keep the output away from the model.
        pendingAsks.delete(key);
        await client.resolve(unresolved.ref, false, "adapter:unresolved-ask").catch(() => undefined);
        return {
          systemMessage: `writ: '${post.tool_name}' ran without a resolved writ approval; output withheld`,
          hookSpecificOutput: { hookEventName: "PostToolUse", updatedToolOutput: WITHHELD_OUTPUT },
        };
      }
      return {};
    }
    dispatched.delete(key);
    const text = outputText(post.tool_response);
    const redact = tracked.decision.decision === "redact";
    try {
      const recorded = await client.complete(tracked.ref, text === undefined ? { ok: true } : { ok: true, output: text });
      if (!redact || text === undefined) return {};
      if (recorded.output === undefined) {
        return {
          systemMessage: "writ: redact verdict but no redacted output was returned; output withheld",
          hookSpecificOutput: { hookEventName: "PostToolUse", updatedToolOutput: WITHHELD_OUTPUT },
        };
      }
      return {
        hookSpecificOutput: {
          hookEventName: "PostToolUse",
          updatedToolOutput: fromRedacted(recorded.output, post.tool_response),
        },
      };
    } catch (err) {
      const message = `writ: could not record '${post.tool_name}' execution: ${errorText(err)}`;
      if (!redact) return { systemMessage: message };
      return {
        systemMessage: `${message}; output withheld`,
        hookSpecificOutput: { hookEventName: "PostToolUse", updatedToolOutput: WITHHELD_OUTPUT },
      };
    }
  };

  const postToolUseFailure: HookCallback = async (input: HookInput, toolUseID) => {
    if (input.hook_event_name !== "PostToolUseFailure") return {};
    const failed = input as PostToolUseFailureHookInput;
    const key = failed.tool_use_id || toolUseID || "";
    const tracked = dispatched.get(key);
    if (tracked === undefined) return {};
    dispatched.delete(key);
    await client.complete(tracked.ref, { ok: false, output: failed.error }).catch(() => undefined);
    return {};
  };

  const canUseTool: CanUseTool = async (toolName, input, opts): Promise<PermissionResult> => {
    const pending = pendingAsks.get(opts.toolUseID);
    if (pending === undefined) {
      if (options.canUseTool !== undefined) {
        try {
          const res = await options.canUseTool(toolName, input, opts);
          return res ?? { behavior: "deny", message: "writ: permission callback returned no result (fail closed)" };
        } catch (err) {
          return { behavior: "deny", message: `writ: permission callback failed (fail closed): ${errorText(err)}` };
        }
      }
      return { behavior: "deny", message: `writ: no approver for '${toolName}' (fail closed)` };
    }
    pendingAsks.delete(opts.toolUseID);
    let approved = false;
    try {
      const res = options.canUseTool !== undefined && !opts.signal.aborted ? await options.canUseTool(toolName, input, opts) : null;
      // writ decided on the original input; an edited input is not what was approved.
      approved = res?.behavior === "allow" && (res.updatedInput === undefined || sameJson(res.updatedInput, input));
    } catch {
      approved = false;
    }
    try {
      const decision = await client.resolve(pending.ref, approved, "sdk:canUseTool");
      if (!shouldDispatch(decision)) {
        return { behavior: "deny", message: describeBlock({ ...pending.decision, ...decision }, toolName) };
      }
      track(dispatched, opts.toolUseID, { ref: decision.ref ?? pending.ref, decision });
      // No updatedPermissions: a persistent SDK allow rule would bypass future writ asks.
      return { behavior: "allow", updatedInput: input };
    } catch (err) {
      return { behavior: "deny", message: `writ: cannot resolve approval for '${toolName}' (fail closed): ${errorText(err)}` };
    }
  };

  const matcher = (hook: HookCallback): HookCallbackMatcher => {
    const m: HookCallbackMatcher = { hooks: [hook] };
    if (options.matcher !== undefined) m.matcher = options.matcher;
    if (options.hookTimeoutSec !== undefined) m.timeout = options.hookTimeoutSec;
    return m;
  };

  return {
    hooks: {
      PreToolUse: [matcher(preToolUse)],
      PostToolUse: [matcher(postToolUse)],
      PostToolUseFailure: [matcher(postToolUseFailure)],
    },
    canUseTool,
    preToolUse,
    postToolUse,
    postToolUseFailure,
  };
}

/** Just the `hooks` option. Deferred asks need `approver` (or use `createWritIntegration` for `canUseTool`). */
export function writHooks(options: WritClaudeOptions): Partial<Record<HookEvent, HookCallbackMatcher[]>> {
  return createWritIntegration(options).hooks;
}

/**
 * Merge writ's hooks with your own `hooks` option. writ's matchers come first
 * for each event; yours are kept.
 */
export function mergeHooks(
  ...sets: Array<Partial<Record<HookEvent, HookCallbackMatcher[]>> | undefined>
): Partial<Record<HookEvent, HookCallbackMatcher[]>> {
  const out: Partial<Record<HookEvent, HookCallbackMatcher[]>> = {};
  for (const set of sets) {
    if (set === undefined) continue;
    for (const [event, matchers] of Object.entries(set) as Array<[HookEvent, HookCallbackMatcher[] | undefined]>) {
      if (matchers === undefined) continue;
      out[event] = [...(out[event] ?? []), ...matchers];
    }
  }
  return out;
}
